### PR TITLE
Implement #264

### PR DESCRIPTION
**Pull request by dweemx**
Improve robustness of `getMarkerGenes` backend function in `GServer.py`

### CODE REVIEW

File: opt/scopeserver/dataserver/modules/gserver/GServer.py [hunk 1]
```diff
@@ -580,19 +580,57 @@ def getMarkerGenes(self, request, context):
...
-        return s_pb2.MarkerGenesReply(genes=genes, metrics=cluster_marker_metrics)
+        return s_pb2.MarkerGenesReply(
+            genes=cluster_marker_metrics_nonzero.index,
+            metrics=list(
+                map(
+                    to_pb,
+                    md_cmm
+                )
+            )
+        )
```
Comment (jhc-at-home) [inline opt/scopeserver/dataserver/modules/gserver/GServer.py]: I would prefer you assign this an identifier before passing it to `MarkerGenesReply`. `map` (sadly) is generally frowned upon in Python.
```suggestion
        metrics = [to_pb(x) for x in md_cmm]
        return s_pb2.MarkerGenesReply(
            genes=cluster_marker_metrics_nonzero.index,
            metrics=metrics
        )
```

File: opt/scopeserver/dataserver/modules/gserver/GServer.py [hunk 2]
```diff
@@ -580,19 +580,57 @@ def getMarkerGenes(self, request, context):
...
             def create_cluster_marker_metric(metric):
-                cluster_marker_metrics = loom.get_cluster_marker_metrics(
-                    clustering_id=request.clusteringID, cluster_id=request.clusterID, metric_accessor=metric["accessor"]
+                return loom.get_cluster_marker_metrics(
+                    clustering_id=request.clusteringID,
+                    cluster_id=request.clusterID,
+                    metric_accessor=metric["accessor"]
                 )
+
+            def to_pb(metric):
```
Comment (jhc-at-home) [inline opt/scopeserver/dataserver/modules/gserver/GServer.py]: Can you think of a clearer name than `to_pb`?

File: opt/scopeserver/dataserver/modules/gserver/GServer.py [hunk 3]
```diff
@@ -580,19 +580,57 @@ def getMarkerGenes(self, request, context):
...
+                    left_index=True,
+                    right_index=True,
+                    how='outer'
+                ),
+                list(
+                    map(
+                        lambda x: create_cluster_marker_metric(x),
+                        md_cmm
+                    )
+                )
```
Comment (jhc-at-home) [inline opt/scopeserver/dataserver/modules/gserver/GServer.py]:
```suggestion
                [create_cluster_marker_metric(x) for x in md_cmm]
```

File: opt/scopeserver/dataserver/modules/gserver/GServer.py [hunk 4]
```diff
@@ -580,19 +580,57 @@ def getMarkerGenes(self, request, context):
...
                 return s_pb2.MarkerGenesMetric(
                     accessor=metric["accessor"],
                     name=metric["name"],
                     description=metric["description"],
-                    values=cluster_marker_metrics,
+                    values=cluster_marker_metrics[metric["accessor"]]
                 )
 
-            cluster_marker_metrics = list(map(create_cluster_marker_metric, md_cmm))
+            cluster_marker_metrics = functools.reduce(
```
Comment (jhc-at-home) [inline opt/scopeserver/dataserver/modules/gserver/GServer.py]: This deserves to be pulled into a new function

File: opt/scopeserver/dataserver/modules/gserver/GServer.py (hunk 2 above)
Comment (dweemx) [reply opt/scopeserver/dataserver/modules/gserver/GServer.py]: Following @KrisDavie style, I could use `protoize_cluster_marker_metric` ; Does that sound good ?

File: opt/scopeserver/dataserver/modules/gserver/GServer.py (hunk 1 above)
Comment (dweemx) [reply opt/scopeserver/dataserver/modules/gserver/GServer.py]: I'm happy to hear best practices in Python, also eager to know why ? :)

File: opt/scopeserver/dataserver/modules/gserver/GServer.py (hunk 2 above)
Comment (jhc-at-home) [reply opt/scopeserver/dataserver/modules/gserver/GServer.py]: Yes I prefer that

File: opt/scopeserver/dataserver/modules/gserver/GServer.py (hunk 1 above)
Comment (jhc-at-home) [reply opt/scopeserver/dataserver/modules/gserver/GServer.py]: A blog post by Guido is a good start: https://www.artima.com/weblogs/viewpost.jsp?thread=98196
Also top comment by Guido on PEP 279: https://www.python.org/dev/peps/pep-0279/

Comment (dweemx) [reply opt/scopeserver/dataserver/modules/gserver/GServer.py]: Great! thanks for the links :+1:

File: opt/scopeserver/dataserver/modules/gserver/s_pb2.py [hunk 5]
```diff
@@ -1,6949 +1,4367 @@
 # -*- coding: utf-8 -*-
 # Generated by the protocol buffer compiler.  DO NOT EDIT!
 # source: s.proto
-# pylint: skip-file
-# type: ignore
```
Comment (KrisDavie) [inline opt/scopeserver/dataserver/modules/gserver/s_pb2.py]: These should stay to make the linting happier. We should alter the bash script generating these to add them back in and then black the outputs

File: opt/scopeserver/dataserver/modules/gserver/s_pb2_grpc.py [hunk 6]
```diff
@@ -1,454 +1,454 @@
 # Generated by the gRPC Python protocol compiler plugin. DO NOT EDIT!
-# pylint: skip-file
```
Comment (KrisDavie) [inline opt/scopeserver/dataserver/modules/gserver/s_pb2_grpc.py]: Same here as on s_pb2

File: opt/scopeserver/dataserver/modules/gserver/s_pb2.py (hunk 5 above)
Comment (dweemx) [reply opt/scopeserver/dataserver/modules/gserver/s_pb2.py]: Merged master into this branch ;)

File: opt/scopeserver/dataserver/modules/gserver/s_pb2_grpc.py (hunk 6 above)
Comment (dweemx) [reply opt/scopeserver/dataserver/modules/gserver/s_pb2_grpc.py]: Merged master into this branch ;)

File: opt/scopeserver/dataserver/modules/gserver/s_pb2.py (hunk 5 above)
Comment (dweemx) [reply opt/scopeserver/dataserver/modules/gserver/s_pb2.py]: Weird, still had to still correct some files...

File: opt/scopeserver/dataserver/modules/gserver/s_pb2_grpc.py (hunk 6 above)
Comment (dweemx) [reply opt/scopeserver/dataserver/modules/gserver/s_pb2_grpc.py]: Weird, still had to still correct some files...

File: opt/scopeserver/dataserver/modules/gserver/GServer.py [hunk 7]
```diff
@@ -581,56 +581,33 @@ def getMarkerGenes(self, request, context):
...
                 )
 
-            def to_pb(metric):
+            def protoize_cluster_marker_metric(metric):
                 return s_pb2.MarkerGenesMetric(
                     accessor=metric["accessor"],
                     name=metric["name"],
                     description=metric["description"],
-                    values=cluster_marker_metrics[metric["accessor"]]
+                    values=cluster_marker_metrics[metric["accessor"]],
```
Comment (jhc-at-home) [inline opt/scopeserver/dataserver/modules/gserver/GServer.py]: Why the trailing comma?

Comment (dweemx) [reply opt/scopeserver/dataserver/modules/gserver/GServer.py]: I don't know seems that black is adding it...

Comment (KrisDavie) [reply opt/scopeserver/dataserver/modules/gserver/GServer.py]: https://test-black.readthedocs.io/en/style-guide/style_guide/trailing_commas.html

**Pull request merged**